File: web/lib/folders.ts
// Folders + share client — /api/v1/folders, the per-conversation share
// endpoints, and the public /api/v1/share/{share_id} (deliberately
// unauthenticated so shared links work for anyone).

import { api } from "./api";
import type {
  Folder,
  SharedConversation,
  ShareResult,
} from "./types";

// ------------------------------------------------------------------ folders

export function listFolders(signal?: AbortSignal): Promise<Folder[]> {
  return api.get<Folder[]>("/api/v1/folders", signal);
}

export function createFolder(name: string): Promise<Folder> {
  return api.post<Folder>("/api/v1/folders", { name });
}

export function renameFolder(id: string, name: string): Promise<Folder> {
  return api.patch<Folder>(`/api/v1/folders/${id}`, { name });
}

export function deleteFolder(id: string): Promise<{ ok: boolean }> {
  return api.delete<{ ok: boolean }>(`/api/v1/folders/${id}`);
}

// -------------------------------------------------------------------- share

export function shareConversation(conversationId: string): Promise<ShareResult> {
  return api.post<ShareResult>(`/api/v1/conversations/${conversationId}/share`);
}

export function unshareConversation(conversationId: string): Promise<{ ok: boolean }> {
  return api.delete<{ ok: boolean }>(`/api/v1/conversations/${conversationId}/share`);
}

/** Public share page fetch — no session, no auth header. */
export function getSharedConversation(
  shareId: string,
  signal?: AbortSignal,
): Promise<SharedConversation> {
  return api.get<SharedConversation>(`/api/v1/share/${shareId}`, signal);
}

File: api/jyoti_api/api/folders.py
"""Folders + share endpoints: /api/v1/folders/*, /api/v1/share/*"""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter
from pydantic import BaseModel, Field

from jyoti_api.api.deps import CurrentUser, SessionDep
from jyoti_api.config import get_settings
from jyoti_api.domain import conversations, folders

router = APIRouter(prefix="/api/v1", tags=["folders"])


class FolderCreate(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    parent_id: str | None = None


class FolderUpdate(BaseModel):
    name: str = Field(min_length=1, max_length=120)


# ------------------------------------------------------------------ folders


@router.get("/folders")
def list_folders(user: CurrentUser, session: SessionDep) -> list[dict[str, Any]]:
    return folders.list_folders(session, user.id)


@router.post("/folders", status_code=201)
def create_folder(body: FolderCreate, user: CurrentUser, session: SessionDep) -> dict[str, Any]:
    folder = folders.create_folder(session, user.id, body.name, body.parent_id)
    return folders.to_folder_dto(session, folder)


@router.patch("/folders/{folder_id}")
def update_folder(
    folder_id: str, body: FolderUpdate, user: CurrentUser, session: SessionDep
) -> dict[str, Any]:
    folder = folders.rename_folder(session, user.id, folder_id, body.name)
    return folders.to_folder_dto(session, folder)


@router.delete("/folders/{folder_id}")
def delete_folder(folder_id: str, user: CurrentUser, session: SessionDep) -> dict[str, bool]:
    folders.delete_folder(session, user.id, folder_id)
    return {"ok": True}


# -------------------------------------------------------------------- share


@router.post("/conversations/{conversation_id}/share")
def share_conversation(
    conversation_id: str, user: CurrentUser, session: SessionDep
) -> dict[str, Any]:
    conversation = conversations.create_share(session, user.id, conversation_id)
    settings = get_settings()
    url = f"{settings.webui_url.rstrip('/')}/s/{conversation.share_id}"
    return {"share_id": conversation.share_id, "url": url}


@router.delete("/conversations/{conversation_id}/share")
def unshare_conversation(
    conversation_id: str, user: CurrentUser, session: SessionDep
) -> dict[str, bool]:
    conversations.clear_share(session, user.id, conversation_id)
    return {"ok": True}


@router.get("/share/{share_id}")
def get_shared_conversation(share_id: str, session: SessionDep) -> dict[str, Any]:
    """Public share link — deliberately outside the auth dependency."""
    return conversations.get_public_share(session, share_id)
